addons/project_timesheet_holidays/tests/test_timesheet_global_time_off.py

```diff
@@ -110,3 +110,39 @@ def test_no_timesheet_on_off_days(self):
         # No timesheet should have been created on the day off
         timesheet = self.env['account.analytic.line'].search([('date', '=', day_off), ('task_id', '=', leave_task.id)])
         self.assertFalse(timesheet.id)
+
+    # This tests that timesheets are created/deleted for every employee with the same calendar
+    # when a global time off has a calendar_id set/remove
+    def test_timesheet_creation_and_deletion_for_calendar_set_and_remove(self):
+        leave_start_datetime = datetime(2021, 1, 4, 7, 0, 0, 0)  # This is a monday
+        leave_end_datetime = datetime(2021, 1, 8, 18, 0, 0, 0)  # This is a friday
+
+        global_time_off = self.env['resource.calendar.leaves'].create({
+            'name': 'Test',
+            'calendar_id': self.test_company.resource_calendar_id.id,
+            'date_from': leave_start_datetime,
+            'date_to': leave_end_datetime,
+        })
+
+        # 5 Timesheets should have been created for full_time_employee and full_time_employee_2
+        # but none for part_time_employee
+        leave_task = self.test_company.leave_timesheet_task_id
+
+        # Now we delete the calendar_id. The timesheets should be deleted too.
+        global_time_off.calendar_id = False
+
+        self.assertFalse(leave_task.timesheet_ids.ids)
+
+        # Now we reset the calendar_id. The timesheets should be created and have the right value.
+        global_time_off.calendar_id = self.test_company.resource_calendar_id.id
+
+        timesheets_by_employee = defaultdict(lambda: self.env['account.analytic.line'])
+        for timesheet in leave_task.timesheet_ids:
+            timesheets_by_employee[timesheet.employee_id] |= timesheet
+        self.assertFalse(timesheets_by_employee.get(self.part_time_employee, False))
+        self.assertEqual(len(timesheets_by_employee.get(self.full_time_employee)), 5)
+        self.assertEqual(len(timesheets_by_employee.get(self.full_time_employee_2)), 5)
```

Comment:
Since you just want to check the number of timesheets per employee, you could have the number of timesheets as value instead of having a recordset of timesheets.
```suggestion
        timesheets_by_employee = defaultdict(int)
        for timesheet in leave_task.timesheet_ids:
            timesheets_by_employee[timesheet.employee_id.id] += 1
        self.assertFalse(timesheets_by_employee.get(self.part_time_employee.id, False))
        self.assertEqual(timesheets_by_employee.get(self.full_time_employee.id), 5)
        self.assertEqual(timesheets_by_employee.get(self.full_time_employee_2.id), 5)
```
Also in the keys of your dict, we could just store the employee id since we don't need the recordset.

Reply:
I did it that way to be consistent with the test above test_timesheet_creation_and_deletion_for_time_off, should I also modify this other test?

Reply:
Oh indeed, I didn't know that and no in stable we prefer to keep a minimal diff. 🙂 
For your test, it is up to you, but I think my suggestion it is less costly than the way done in the previous test. (However I agree it is not huge as improvement)

Reply:
I think for three employee records it's not worth the change in a test but I agree that the change would be a nice performance improvement :)